// f1-22-udp/build/src/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.packetSize = exports.packetID = void 0;
var packetID;
(function (packetID) {
    packetID[packetID["Motion"] = 0] = "Motion";
    packetID[packetID["Session"] = 1] = "Session";
    packetID[packetID["LapData"] = 2] = "LapData";
    packetID[packetID["Event"] = 3] = "Event";
    packetID[packetID["Participants"] = 4] = "Participants";
    packetID[packetID["CarSetups"] = 5] = "CarSetups";
    packetID[packetID["CarTelemetry"] = 6] = "CarTelemetry";
    packetID[packetID["CarStatus"] = 7] = "CarStatus";
    packetID[packetID["FinalClassification"] = 8] = "FinalClassification";
    packetID[packetID["LobbyInfo"] = 9] = "LobbyInfo";
    packetID[packetID["CarDamage"] = 10] = "CarDamage";
    packetID[packetID["SessionHistory"] = 11] = "SessionHistory";
    packetID[packetID["Unknown"] = 12] = "Unknown";
})(packetID = exports.packetID || (exports.packetID = {}));
var packetSize;
(function (packetSize) {
    packetSize[packetSize["Motion"] = 1464] = "Motion";
    packetSize[packetSize["Session"] = 632] = "Session";
    packetSize[packetSize["LapData"] = 972] = "LapData";
    packetSize[packetSize["Event"] = 40] = "Event";
    packetSize[packetSize["Participants"] = 1257] = "Participants";
    packetSize[packetSize["CarSetups"] = 1102] = "CarSetups";
    packetSize[packetSize["CarTelemetry"] = 1347] = "CarTelemetry";
    packetSize[packetSize["CarStatus"] = 1058] = "CarStatus";
    packetSize[packetSize["FinalClassification"] = 1015] = "FinalClassification";
    packetSize[packetSize["LobbyInfo"] = 1191] = "LobbyInfo";
    packetSize[packetSize["CarDamage"] = 948] = "CarDamage";
    packetSize[packetSize["SessionHistory"] = 1155] = "SessionHistory";
})(packetSize = exports.packetSize || (exports.packetSize = {}));
